### ISSUE TITLE
Inserting a file into an index which overwrites a directory silently discards the file when a tree is produced.

### ISSUE THREAD

**Issue by Parakleta** [libgit2/libgit2#4077]
### Reproduction steps

```python
from tempfile import mkdtemp
from pygit2 import init_repository, Index, IndexEntry, GIT_FILEMODE_BLOB

repo = init_repository(mkdtemp(),bare=True)
blob1 = repo.create_blob("data1")
blob2 = repo.create_blob("data2")
blob3 = repo.create_blob("data3")

idx = Index()
idx.add(IndexEntry('afile',blob1,GIT_FILEMODE_BLOB))
idx.add(IndexEntry('pathto/afile',blob2,GIT_FILEMODE_BLOB))
idx.add(IndexEntry('pathto',blob3,GIT_FILEMODE_BLOB))

for entry in idx:
    print(entry.path,entry.id)
```

### Expected behavior

```
('afile', 0abc8f194801d3d07af700bae67026ed2695ec59)
('pathto', cdca2c1e94f3d30ed4e7822b2da91c59792899b4)
```

### Actual behavior

```
('afile', 0abc8f194801d3d07af700bae67026ed2695ec59)
('pathto', cdca2c1e94f3d30ed4e7822b2da91c59792899b4)
('pathto/afile', fee93d14fc11d09aeffc1494c3d8cb5b3635b9f0)
```

### Version of libgit2 (release number or SHA1)

Version 0.24.2 (through pygit2 of the same version)

### Operating system(s) tested

FreeBSD

### Other Information

The bug was introduced in commit 21515f228b739a3ecd2237bafbba50e8d219d8dd.  The mechanism is that previously when `index_insert` was run it would return the insertion position of the new entry from the call to `index_find` even if the entry wasn't found.  This position information is required for `check_file_directory_collision` and more specifically as the starting position for the search in `has_file_name`.  After the changes introduced in that commit the call to `index_existing_and_best` which now wraps the call to `index_find` discards the position information if the exact entry wasn't found.

I also believe that there is a mistake in the behaviour of `has_file_name` and `has_dir_name` in that they both return an error which causes the `index_insert` to fail even if its variable `replace` (renamed `ok_to_replace` in the `has_*_name` functions) is set.  In this situation they both remove the replaced entries but then due to the error the replacement is never inserted.  I think their `ok_to_replace` code should change to something like:

```c
if (!ok_to_replace || index_remove_entry(index, pos) < 0) {
    retval = -1;
    break;
}
```

So that the `retval = -1` is only used if the replacement is not acceptable or it fails.

**Comment by carlosmn** [comment]
That output you report does show a bug, but it looks like a different one from the one you are reporting. It is invalid for an index to have both an entry for `pathto` and `pathto/afile`. It should never get to the point of writing it out to a tree since it should error out when you ask it to insert `pathto`.

**Comment by Parakleta** [comment]
Yes, I should have changed the title.  When I first encountered the bug the issue I had was with the tree that was being output, it was only while writing up the bug report that I realised it was simpler to show how the index was being assembled.

**Comment by carlosmn** [comment]
We've had tests for this since 95fbedcd8eaaff2747196e20abd959596baf7e98 but it actually looks like the behavior is like the opposite right now of what's described here. Inserting `a/b` and then `a/b/c` deletes `a/b`. I'm going to send a PR so we at least see some failing tests here.

**Comment by ethomson** [comment]
> Inserting a/b and then a/b/c deletes a/b. I'm going to send a PR so we at least see some 

This is what I would expect to happen.  🤔
